### ISSUE TITLE
Add documentation and script to update the nuclides.dat file data

### ISSUE THREAD

**Issue by jakehader** [terrapower/armi#1013]
On the review of https://github.com/terrapower/armi/pull/998 there was a request to create more detailed documentation on how the data for `nuclides.dat` was acquired and to provide a script that accesses the RIPL-3 library data and the IAEA chart of the nuclides API to more readily allow a future developer to update these the nuclides in the framework. This is a great request, but would take more time to close out that PR. This issue is related to completing that documentation of the process.

**Comment by john-science** [comment]
@jakehader  This ticket is still assigned to you.  That's cool with me, but if you aren't planning on working on it, please un-assign it.

(You are not being singled out, I am going through all the ARMI tickets.)

**Comment by jakehader** [comment]
Thanks @john-science for the call out! Sounds like a fun weekend activity 😁

**Comment by gorbion** [comment]
Hi, guys! This is my very first contact with ARMI community. I hope in the long run to prove helpful in this oustanding and incredibly interesting project, still, I'm aware that my first steps can be painful and should be taken with care...

Anyway, I hope you wouldn't mind if I took a try at this issue. In fact I've already started some implementation efforts but at the moment I'm facing two questions that are beyond my current rather limited knowledge in nuclear engineering area in general, and ARMI in particular. I plan to provide a  Python module with a bunch of classes responsible for gathering and processing external data, and then creating/updating `nuclides.dat` file. Before I submit PR I'd like to make sure it is of decent completness and quality. Thus, the aforementioned questions:

1. A considerable number of entries in `nuclides.dat` is missing from RIPL3 mass-related data files, **<sup>13</sup>Li** being the first, and **<sup>20</sup>B**, **<sup>21</sup>B** or **<sup>11</sup>O** among others. How can one define the "AMRI full set of nuclides"? What is the source data for such a "full set"?  I hope it won't turn out to be a chicken and egg situation ;-) because among other features I'd like the new module to provide the functionality of creating `nuclides.dat` from scratch which means I need some way of determining initial "AMRI full set of nuclides" (possibly also providing some comfortable way for modifying such "init set").
2. Furthermore, when there is no external mass excess data avaliable how should one find nuclide's amu mass (for **<sup>13</sup>Li**, **<sup>20</sup>B**, etc.)? Quick inspection of RIPL3 materials led me for instance to _Duflo-Zuker mass formula_. Is that the way to go? If so, futrther questions are bound to arise, I'm afraid, but at least the direction is set for me.

If you consider taking a look at my first implementation worth your time, please let me know. I can already open PR in its current state - a bit dirty and definitely incomplete, albeit functional and accompanied with tests. But of course I'm not going to do that without some kind of upfront invitation on your part as it still cannot be consider ready in terms of _Standards and Practises_.

**Comment by keckler** [comment]
Hi @gorbion , thanks for your questions and your interest in contributing. 

I think you might have picked quite a thorny issue to start your ARMI experience with :-) As you've noted, there are a few ambiguous questions with this one. @jakehader would be the one to answer those questions, since he put together the `nuclides.dat` file. 

Speaking frankly, unless @jakehader has some pointers for you, unfortunately I think that this might not be the best issue for a newcomer. If you are interested in contributing to ARMI in other ways, I'd recommend that you take a look at our issues and make a comment on any of them before you begin working to see if it would be suitable. That way you don't get too far before getting stuck in a quagmire.

**Comment by gorbion** [comment]
Hi @keckler, thank you for the response! I guess I fully understand your scepticism about whether I'm the right person to get involved into the task - especially as the first assignment ever... Still, I firmly believe the only way to truly gain experience in something is to get one's hands dirty. The dirtier, the better ;-) (not beyond common sense, of course). And I have tried to do exactly that. Moreover, as I understand nuclides' data lies at the very bottom of what we're dealing with here. As such, its reliability and up-to-dateness is crucial, and automation of external data retrieval and processing seems to be the only simple way to guarantee just that. Basically that's the motivation behind my _dedication_ to this issue.

Now, apart from questions to be asked I also do have some results. Please find attached:
- The updated `nuclides.dat` file (I had to add superfluous `.txt` extension to satisfy GitHub upload rules) that has been generated by the current version of my module (full automation, all resources downloaded on-the-fly). Having no clue as for how to deal with "missing data for some nuclides" (I've recently asked about that) I decided to introduce the mode in which entries for which there is no external data are simply left intact. Such are the contents of the attached file. _Diff_ it against current ARMI `nuclides.dat` to get the full picture.
- `diff_summary.txt` file that contains simplified, "tabulated" summary of the differences betwen the **_old_** and the **_new_** `nuclides.dat` (I've used ARMI's `tabulate` utility - it's such a canny tool!).

I've double checked decent amount of those differences and I believe all ground state related data is correctly updated in full accordance with current state of external sources. At the same time excited/metastable states data is a different story. As you might imagine, I've had some really hard time trying to identify the rules governing _extraction_ of those states from RIPL-3 Discrete Levels and Decay Data library files. I believe the output now seems reasonably reliable, nevertheless thorough expert-level review is needed. No doubt about that.

@jakehader, I would greatly appreciate if you took a look at what has already been said and at the provided material in order to decided whether it all looks like a decent half-product worth "final brush strokes" or I should indeed give it up, and look for some more-novice-friendly taks(s).

[nuclides.dat.txt](https://github.com/user-attachments/files/18137399/nuclides.dat.txt)
[diff_summary.txt](https://github.com/user-attachments/files/18137401/diff_summary.txt)

**Comment by jakehader** [comment]
Hi @gorbion - here is the original RIPL-3 reader / parser that was removed when the nuclide data was updated. While RIPL-3, is a good source for nuclide data, I did notice various inconsistencies across RIPL-3, ENDF-B/V-II.1, TENDL, and other sources that leverage ENDF data for visualization (e.g., chart of nuclides from the IAEA JANIS app, nndc). 

Some of the hopefully nearer term vision for ARMI is to remove the global nuclides state from the system (#473) and to allow users within the application or for their specific use case to configure nuclide data from sources that they wish to use as opposed to enforcing it through a single and difficult to interpret text file (`nuclides.dat`). 

I think for this clean up effort, there is a lack of documentation for how I obtained the data now in `nuclides.dat`, but an audit of the data would be very valuable  (#1085).

First, welcome and thank you for reaching out to help! I think there are many attempts at nuclide definitions out there on various nuclear engineering GitHub projects but if you'd like to catalog different resources available and perform some comparisons independently that may help users in the future decide which library they'd like to use. Additionally, ARMI could even provide a `nuclide` starter kit for single sources (e.g., pure RIPL-3, pure ENDF, pure TENDL, etc.). Right now `nuclides.dat` is an amalgamation of sources where if at least 2 sources agreed on a value that value was taken as "truth". The metastable state and more obscure nuclides are the ones that likely have the most uncertainty and therefore likely vary the most between sources. Documenting the "confidence" in this data could also be pretty important. Most reactor design calculations rely on a few key nuclides, including fission products, so having large uncertainties or low confidence could be deemed acceptable for some applications / use cases, but not all applications / use cases are the same and it should be left up to the end user to determine if the defaults provided by ARMI are truly acceptable.

These are some of my thoughts on this, but since it looks like you're excited to get into it, feel free to post some ideas of where you'd like to start and I'm happy to provide more pointers towards different sources.

**Comment by jakehader** [comment]
> Hi, guys! This is my very first contact with ARMI community. I hope in the long run to prove helpful in this oustanding and incredibly interesting project, still, I'm aware that my first steps can be painful and should be taken with care...
> 
> Anyway, I hope you wouldn't mind if I took a try at this issue. In fact I've already started some implementation efforts but at the moment I'm facing two questions that are beyond my current rather limited knowledge in nuclear engineering area in general, and ARMI in particular. I plan to provide a Python module with a bunch of classes responsible for gathering and processing external data, and then creating/updating `nuclides.dat` file. Before I submit PR I'd like to make sure it is of decent completness and quality. Thus, the aforementioned questions:
> 
> 1. A considerable number of entries in `nuclides.dat` is missing from RIPL3 mass-related data files, **13Li** being the first, and **20B**, **21B** or **11O** among others. How can one define the "AMRI full set of nuclides"? What is the source data for such a "full set"?  I hope it won't turn out to be a chicken and egg situation ;-) because among other features I'd like the new module to provide the functionality of creating `nuclides.dat` from scratch which means I need some way of determining initial "AMRI full set of nuclides" (possibly also providing some comfortable way for modifying such "init set").
> 2. Furthermore, when there is no external mass excess data avaliable how should one find nuclide's amu mass (for **13Li**, **20B**, etc.)? Quick inspection of RIPL3 materials led me for instance to _Duflo-Zuker mass formula_. Is that the way to go? If so, futrther questions are bound to arise, I'm afraid, but at least the direction is set for me.
> 
> If you consider taking a look at my first implementation worth your time, please let me know. I can already open PR in its current state - a bit dirty and definitely incomplete, albeit functional and accompanied with tests. But of course I'm not going to do that without some kind of upfront invitation on your part as it still cannot be consider ready in terms of _Standards and Practises_.

@gorbion - The full set of nuclides in ARMI is definitely an ambiguous definition. My previous comment mentions use cases, so in this case the set available aligns with our use cases. For instance, we care very little about B-20 since it is very far from the island of stability, it does not exist in typical nuclear reactor applications since transmutation or decay to B-20 would be extremely rare or unquantifiable. 

I like your idea of exposing data parsers / readers or having them available for the community (see above about global nuclides), but I'm not exactly sure the best way to maintain them. Out of curiosity, are you working with ARMI to make an `app` where any differences you noticed are important?

**Comment by gorbion** [comment]
@jakehader Thanks for your inavaluable feedback! It turns out my orignal viepoint was seriously limited. I took for granted that there exists something like "ARMI full set of nuclides". As if there could exist such thing as "ARMI-centric nuclides model"... Obviously now I understand there can't, rightly so. Let me rephrase basic idea. ARMI can (even should) provide some standardized "models" in this regard (e.g. RIPL-based, ENDF-based, etc.) but in general the user should be free to (re)define their own "model" that drives their needs. I thought that the current nuclide model's "global" characteristics had been introduced "by design". I only now see it's just the other way round - _globality_ has to be get rid of. ASAP. Hope I get it all right now.

Anyway, while "digging" around #1013 I somehow missed #473 and problems referenced therein. I'd love to help here, however I'm not sure if I can be productive enough right away (without prior experience in ARMI). One thing is to manage and serve the models in efficient/convinient way but another is to get and prepare (pre-process?) reliable models in the first place. So instead how about if I made an attempt at preparing some generalized _abstract nuclides model_ along with, say, three first implementations, namely "legacy ARMI", "pure RIPL-3", and "pure ENDF"? Standardized interface, specialized create/refresh procedures, etc. For the time being I have no idea how to deal with "certainty metrics" or "error levels" or whatever we call among-models inconsistencies in the long run. Some issues undoubtedly up for discussion/decision, I definitely need guidance here but I believe this is almost exactly what you've had in mind in your comments. And I could start practically right away.

Furthermore, such models could potentially be directly utilized by means of some _strategy_ design pattern to dynamically load on-the-fly in ARMI applications' _local_ scope. Full, flexible customization by end users is another, more complcated story but at least we're taking it into account from the very beginning. Along the way I should also take care of #1085 (at least to some extent as my current limited level of expertise in nuclear engineering nitty-gritty details and unevitable traps can surely become evident...)

Ideally, after I finish what is described above, you can directly use the abstract nuclides model and its first implementations in #473. Alternatively, if by that time #473 will already have its own model in place, I should at least be able to quickly make necesary adaptations (I definitely opt for preparing some flexible, loosely-coupled components) in order to provide #473 with ready-to-use model _incarnations_.

What do you think? If all seems reasonable (or even needs some serious refinement but otherwise makes perfect sense) I'd love to confirm the details of the first steps. And get going :-)
